OCF module v2

rafalste opened the pull request:
Change-Id: I58f0fa973cc78b1c7bfefd9ffd783a6f752baba3

Commit:
OCF module v2

Change-Id: I58f0fa973cc78b1c7bfefd9ffd783a6f752baba3
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Remove old OCF module

Change-Id: Icf9704d338fe8fe5d3f6ef8eb5f9255af4a649d2
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

mmichal10 commented on module/bdev/ocf/vbdev_ocf_cache.c:
```diff
@@ -0,0 +1,308 @@
+#include "ctx.h"
+#include "utils.h"
+
+struct vbdev_ocf_caches_head g_vbdev_ocf_caches = STAILQ_HEAD_INITIALIZER(g_vbdev_ocf_caches);
```
Do we really need to maintain a separate list of caches in the adapter? Why not take advantage of OCF's list? `struct vbdev_ocf_cache` could be stored in `cache->priv`

rafalste replied:
tl;dr - I don't know yet...

I started with the approach that all of the other SPDK bdevs (and vbdevs especially) take, by reading a few other modules and getting what I thought was best for our scenario, using tools provided by SPDK and making it "the SPDK way" to comply with their flow as much as possible.
Then during the implementation of OCF into the rest of the code I was thinking about moving some parts (e.g. cache/core lists) into the OCF as well. But there might be lots of complication with this approach. First of all, there are many different states that each device can be in. I'm mostly thinking about the examine process which discovers devices already added by config but not started yet due to some lack of other resources (base devices mostly). I'm working on this examine process and all of its scenarios at the moment and there will be more clarity on your question when it will be finished. There will definitely be a need to store caches/cores on some temporary lists during different stages of this examine process, but if "temporary" is enough and if it's worth switching from one list to another and how often it should happen, all of those things will be more clear when the module will be more complete.
Another things are the SPDK app shutdown process and the registered (exposed) vbdev handling and all of their stages...
During the further development, when the adapter will grow with features and connections between those stages and components I will definitely try to use as much of OCF capabilities as possible, not compromising the SPDK flow and readability at the same time.

mmichal10 replied:
When the base device for cache is missing doesn't it mean that we could simply start cache in detached state (which we will fully support from the upcomming? And once the bdev appears the examine could initiate ocf_cache_attach(). Are there any more states that the adapter should support besides from attached and detached?

mmichal10 commented on module/bdev/ocf/vbdev_ocf_cache.c:
```diff
@@ -0,0 +1,308 @@
+		return -ENOMEM;
+	}
+
+	STAILQ_INIT(&cache->cores);
```
Isn't a separate list for cores redundant as well? OCF already tracks cores for each cache

Commit:
OCF module: move data to OCF structs

Change-Id: Ie8d756c4374f5fb75e39a885b203a6b6883d1bf5
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
fix memleaks

Change-Id: I82aae275e79444afbda292bd0cc8211d18fd5c0e
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Implement bdev examine and hotremove

Change-Id: I74ef4446df4f51d662b31e2774a8d1c7f1c11df0
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Update OCF to 25.03

Change-Id: I6cd3262aed2b309929df68d24fac5b431ecfc006
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Implement cache load

Change-Id: Ie509dc0f48e76cd9c01cb7e07ef8698dfa3bab14
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Management tests

Change-Id: I91828704921fb74ff538ba37e6f6a5a8cbd218bf
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Multiple fixes

- check cache/core block size
- merge cache stop paths
- unify adding cores from waitlist
- error handling
- comments
- cleanup

Change-Id: Ib111089a9e5fe1ec155487db3b2c4541d1ce9520
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Parameterize tests

Change-Id: I1d4ecf9555f9996d236c4985ce460a8839b02e2f
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Implement configuration change RPCs

Change-Id: I0049cda213a5d037819b877cd1279286a1a68841
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Configuration change tests

Change-Id: Ia6aa94a2f352d1f9178958dd34e7e069ea2bc3d9
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Various improvements

Change-Id: Ib36ae38a03d1bca9b2595a78b7cc162ab787ebe8
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Implement statistics

Change-Id: I70e3d6470dc3f0c7a6a493472746ccdc3bd3c6e7
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

mmichal10 commented on module/bdev/ocf/vbdev_ocf_core.c:
```diff
@@ -0,0 +1,501 @@
+		SPDK_ERRLOG("OCF vbdev '%s': failed to create IO queue poller\n", vbdev_name);
+		spdk_put_io_channel(ch_ctx->core_ch);
+		spdk_put_io_channel(ch_ctx->cache_ch);
+		ocf_queue_put(ch_ctx->queue);
```
spdk_put_io_channel() will be also called as part of ocf_queue_put(), so it's a double free.

rafalste replied:
fixed

mmichal10 commented on module/bdev/ocf/vbdev_ocf.c:
```diff
@@ -1,1794 +1,2833 @@
 
-	return status;
+	/* Increment queue refcount to prevent destroying management queue after cache device detach. */
+	ocf_queue_get(((struct vbdev_ocf_cache *)ocf_cache_get_priv(cache))->cache_mngt_q);
```
I think the mngt queue refcnt should be incremented once, on cache start. Incrementing it here feels like a workaround

rafalste replied:
The problem turned out to be in the context cleaner refcount. Fixed.

mmichal10 commented on module/bdev/ocf/vbdev_ocf.c:
```diff
@@ -1,1794 +1,2833 @@
-static int
-vbdev_ocf_init(void)
+static void
+_cache_detach_rpc_detach_cb(ocf_cache_t cache, void *cb_arg, int error)
```
At this point the cache device has been detached but the vbdev io channels which had been created before the detach still keep a pointer to the detached bdev's io channel. When the cache is detached it's just a matter of cleanup, because no IO would be submitted to the cache anyways

rafalste replied:
fixed

mmichal10 commented on module/bdev/ocf/vbdev_ocf.c:
```diff
@@ -1,1794 +1,2833 @@
 static void
-base_bdev_event_cb(enum spdk_bdev_event_type type, struct spdk_bdev *bdev,
-		   void *event_ctx)
+_cache_attach_rpc_attach_cb(ocf_cache_t cache, void *cb_arg, int error)
```
After the new cache bdev has been attached you really need to update the existing vbdev io channels. Otherwise the IO to cache would be written either to the io channel of the previous cache bedev or to uninitialized iochannel.

rafalste replied:
fixed

mmichal10 commented on module/bdev/ocf/vbdev_ocf_core.c:
```diff
@@ -0,0 +1,501 @@
+	SPDK_DEBUGLOG(vbdev_ocf, "OCF vbdev '%s': destroying IO channel\n",
+		      spdk_bdev_get_name(&(((struct vbdev_ocf_core *)ocf_core_get_priv(core))->ocf_vbdev)));
+
+	ocf_queue_put(ch_destroy_ctx->queue);
```
The core_bdev iochannel, the cache_bdev iochannel, the poller and the queue are all owned by vbdev_iochannel. I think you should decrement all their refcounter here on vbdev_channel_destroy just as you're creating them on vbdev_channel_create.

rafalste replied:
Since the channel context with all of those you mention is saved in queue priv, this decrementing will be done in vbdev_ocf_core_io_queue_stop() when the queue is destroyed.

mmichal10 commented on module/bdev/ocf/volume.c:
```diff
@@ -339,6 +344,26 @@ static struct ocf_volume_properties vbdev_volume_props = {
+vbdev_ocf_base_detach(struct vbdev_ocf_base *base)
+{
+	if (base->thread && base->thread != spdk_get_thread()) {
+		spdk_thread_send_msg(base->thread, _base_detach, base);
```
The error returned by `spdk_thread_send_msg()` shouldn't be ignored

rafalste replied:
Almost all usage of spdk_thread_send_msg() in all other modules look like this. But you're right - why not use it if it's there. Fixed.

mmichal10 commented on module/bdev/ocf/ctx.c:
```diff
@@ -285,18 +287,19 @@ vbdev_ocf_queue_put(ocf_queue_t queue)
+	ocf_queue_t queue = ctx;
+	int i, queue_runs;
+
+	queue_runs = spdk_min(ocf_queue_pending_io(queue), VBDEV_OCF_QUEUE_RUN_MAX);
```
Why is there a limit of queue runs?

rafalste replied:
To not run the queue for to long and let it be rescheduled if needed.

mmichal10 commented on module/bdev/ocf/vbdev_ocf.c:
```diff
@@ -1,1794 +1,2833 @@
-		return;
+	SPDK_DEBUGLOG(vbdev_ocf, "OCF cache '%s': module stop visit\n", ocf_cache_get_name(cache));
+
+	/* Increment cache refcount to not destroy cache structs before destroying all cores. */
```
This sounds like a workaround. I think we should have a proper management of cache lifetime.

rafalste replied:
Already removed this bit because it is actually not needed anymore since cache stop is not triggered asynchronously with cores detach any more.
I found a better solution for this in later commits and it is taken care of by this check:
`ocf_cache_get_core_count(cache) == ocf_cache_get_core_inactive_count(cache)`

mmichal10 commented on module/bdev/ocf/vbdev_ocf.c:
```diff
@@ -1,1794 +1,2833 @@
-	if (rc) {
-		SPDK_ERRLOG("Unable to create mngt_queue: %d\n", rc);
-		return rc;
+	// impossible to be true ?
+	if (!core) {
+		SPDK_ERRLOG("OCF vbdev '%s': failed to submit IO - no OCF core device\n",
+			    spdk_bdev_get_name(bdev_io->bdev));
+		spdk_bdev_io_complete(bdev_io, SPDK_BDEV_IO_STATUS_FAILED);
+		return;
 	}
```
If it's impossible it could be a BUG_ON

rafalste replied:
This situation is not possible to occur, so I removed this check. Nevertheless, added an assert here just to be sure.

Commit:
Multiple fixes

- update cache IO channels after device detach/attach
- fix management queue refcount
- fix IO channels double free
- handle spdk_thread_send_msg() errors
- do not allow using OCF core as base bdev for another OCF core
- remove redundant checks
- remove cluttering logs

Change-Id: I34f2aceb434d5752c787ae6071cce3930bc44575
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Implement info ops + tests

- current module config
- generic bdev info
- generic bdev stats
- OCF stats

Change-Id: If63774a399902caaa50c09ef2e89a9837e5db95a
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Address all of the remaining in-code comments

Change-Id: I106829e59cdf145ed62998c6b101d4efd7389821
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Fix formatting

Change-Id: Ib43ab22b3fa457e624deae912b3c587bdf6b0648
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Update documentation

Change-Id: Id6ddfa4042427715f7927c7d66fb98ee9904881c
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Prevent running any calls when module is stopping

Change-Id: I7242902219bcb5604df51f8a06feb50e86426541
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Generate UUIDs for registered bdevs

Change-Id: I3753edb1643a4465c9768e6a8a13cd4c21cc6443
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Refactor new tests

Change-Id: I107349f86aef598b5ec112d1a6df5de20ea4a2c3
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Add integrity tests

Change-Id: Ie01d40cebf469df6d84f5689b2767f6aa7f3721c
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Remove old tests

Change-Id: I488ec6c802299aba39f38743cad12a429ece066e
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Do not print uninitialized core info

Change-Id: Ia187bee78d1e1078eefcb49ebcaf73ffd601f479
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Read trylock cache in bdev_ocf_get_bdevs

Change-Id: I915d5af1cce0ce38bca45462a7b66d14ce8be8e7
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Fix race condition when adding cores in examine

Change-Id: I3607fd8a261755440fdd2948a027a37aaf66d67e
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Move cache IO channel update before attaching new device to OCF

Change-Id: If00f029f084e5f500af7d7e6524cf33193fa1860
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Make sure selected core still exists after taking a cache lock

Change-Id: Ib038d093689c45fade8515861b50cb2b30f5601d
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Add stress tests

Change-Id: I323302f657e0d943acd3b4d4e4c8bd3efd4931d0
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Make a copy of core context when dumping core info

Change-Id: Iaeaf9a9af3c82d908efddf5947a55bf8b2f0fcc2
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Make a copy of bdev name given in RPC call

Change-Id: Ibba93fc9803babcc9c3055aeb223825c580f7743
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Add dirty_ratio_inertia cleaning param

Change-Id: I80e5bc492f9025a13ba4cdc7ca12676bb0ef3f8f
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>

Commit:
Update OCF

Change-Id: I03748efcc653d49fe561731046ff236fcd37314a
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@huawei.com>
